src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -138,124 +198,140 @@ void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_i
     GVTRACE (" new %p", tlm);
   }
-  if (src_ep_guid != NULL)
-  {
-    bool found = false;
-    for (uint32_t n = 0; !found && n < tlm->n_endpoints; n++)
-    {
-      if (!memcmp (&tlm->endpoints[n], src_ep_guid, sizeof (*src_ep_guid)))
-        found = true;
-    }
-    if (!found)
-    {
-      tlm->n_endpoints++;
-      tlm->endpoints = ddsrt_realloc (tlm->endpoints, tlm->n_endpoints * sizeof (*tlm->endpoints));
-      memcpy (&tlm->endpoints[tlm->n_endpoints - 1], src_ep_guid, sizeof (*src_ep_guid));
-      GVTRACE (" add ep "PGUIDFMT, PGUID (*src_ep_guid));
-    }
-  }
+  tlm_endpoint_add (gv, tlm, proxy_ep_guid);
   if (tlm->sertype == NULL && type != NULL)
   {
     tlm->sertype = ddsi_sertype_ref (type);
     tlm->state = TL_META_RESOLVED;
     GVTRACE (" resolved");
+    resolved = true;
   }
   if (dst != NULL && guid_prefix_empty (&tlm->dst_prefix) && !guid_prefix_empty (dst))
     memcpy (&tlm->dst_prefix, dst, sizeof (*dst));
   tlm->refc++;
-  GVTRACE (" state %d refc %u\n", tlm->state, tlm->refc);
+  GVTRACE (" state %d refc %u", tlm->state, tlm->refc);
   ddsrt_mutex_unlock (&gv->tl_admin_lock);
+
+  if (resolved)
+  {
+    ddsrt_mutex_lock (&gv->tl_resolved_lock);
+    ddsrt_cond_broadcast (&gv->tl_resolved_cond);
+    ddsrt_mutex_unlock (&gv->tl_resolved_lock);
+  }
+
   if (type_id == NULL)
     ddsrt_free (tid);
 }
 
-void ddsi_tl_meta_unref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const ddsi_guid_t *ep_guid)
+void ddsi_tl_meta_unref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const struct ddsi_sertype *type, const ddsi_guid_t *proxy_ep_guid)
 {
+  assert (type_id || type);
+  GVTRACE ("unref tl_meta");
+  type_identifier_t *tid;
+  if (type_id != NULL)
+    tid = (type_identifier_t *) type_id;
+  else
+  {
+    GVTRACE (" sertype %p", type);
+    tid = ddsi_typeid_from_sertype (type);
+  }
+  GVTRACE (" tid "PTYPEIDFMT, PTYPEID (*tid));
+
   ddsrt_mutex_lock (&gv->tl_admin_lock);
-  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, tid);
   assert (tlm != NULL);
-  ddsi_tl_meta_unref_impl_locked (gv, tlm, ep_guid);
+  ddsi_tl_meta_unref_impl_locked (gv, tlm, proxy_ep_guid);
   ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  if (type_id == NULL)
+    ddsrt_free (tid);
 }
 
-static struct participant *get_typelookup_pp (const struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix)
+static struct writer *get_typelookup_writer (const struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix, uint32_t wr_eid)
 {
-  struct entidx_enum_participant est;
-  struct participant *pp = NULL;
-  bool pp_found = false;
-  entidx_enum_participant_init (&est, gv->entity_index);
-  while (!pp_found && (pp = entidx_enum_participant_next (&est)) != NULL)
+  struct participant *pp;
+  struct writer *wr = NULL;
+  thread_state_awake (lookup_thread_state (), gv);
+  if (guid_prefix_empty (guid_prefix))
```

Comment:
See the comment about `dst_prefix` where the containing struct type is defined. I'm not so sure it's worth having.